### ISSUE TITLE
efficiency and conservation of boundedness correction

### ISSUE THREAD

**Issue by gforney** [firemodels/fds#932]
```
Handle stretched grids and check efficiency for boundedness correction.
```

Original issue reported on code.google.com by `randy.mcdermott` on 2009-12-07 21:29:56

**Comment by gforney** [comment]
```
We have added the cell volume correction for CHECK_DENSITY and the efficiency does
not seem to be a problem.

We now need to add the correction to CHECK_MASS_FRACTION before we can close this issue.

Jason, in the current version of CHECK_MASS_FRACTION there is a comment bang (!) just
after a THEN so that we have THEN!.  My editor does not recognize this.  Is there any
chance the compiler doesn't like it either?  If we can get rid of the DEL_RHO... <
0._EB check, then let's do it when we update the routine for the cell volume
correction.  Can you make the modifications?  I would do it, but you said there was
something about scalar gradient reversals that you also do in this routine and I do
not want to disturb this.  Thanks.
```

Original issue reported on code.google.com by `randy.mcdermott` on 2010-01-07 14:27:43

**Comment by gforney** [comment]
```
Kevin,

With your new mods to CHECK_MASS_FRACTION are we done with this Issue?

Thanks.
R
```

Original issue reported on code.google.com by `randy.mcdermott` on 2011-12-29 14:17:12

**Comment by gforney** [comment]
```
Both CHECK_DENSITY and CHECK_MASS_FRACTION now account for changes in cell volume due
to grid stretching. Both algorithms have also been streamlined, but they are both essentially
working the same way as before, with the exception of some removal of logic to account
for sign of the mass flux terms and boundary values of mass fraction. It is important
to note that neither routines guarantee that the density and mass fraction will be
restored to within bounds. CHECK_DENSITY does a simple clip at the end to ensure that
density does not go below or above, but CHECK_MASS_FRACTION does not because we want
to maintain species mass conservation.

So let's close this issue for now (I verified that things are working with the V&V
cases), and reopen a new issue if we want to further refine the algorithm.
```

Original issue reported on code.google.com by `mcgratta` on 2011-12-29 14:45:10
